uaclient/log.py

```diff
@@ -9,3 +12,49 @@ class RedactionFilter(logging.Filter):
     def filter(self, record: logging.LogRecord):
         record.msg = util.redact_sensitive_logs(str(record.msg))
         return True
+
+
+class JsonArrayFormatter(logging.Formatter):
+    """Json Array Formatter for our logging mechanism
+    Custom made for Pro logging needs
+    """
+
+    default_time_format = "%Y-%m-%dT%H:%M:%S%z"
+    required_fields = (
+        "asctime",
+        "levelname",
+        "name",
+        "funcName",
+        "lineno",
+        "message",
+    )
+
+    def format(self, record: logging.LogRecord) -> str:
+        record.message = record.getMessage()
+        record.asctime = self.formatTime(record, self.default_time_format)
+        record.asctime = "%s.%03d" % (record.asctime, record.msecs)
+
+        extra_message_dict = {}  # type: Dict[str, Any]
+        if record.exc_info:
+            extra_message_dict["exc_info"] = self.formatException(
+                record.exc_info
+            )
+        if not extra_message_dict.get("exc_info") and record.exc_text:
+            extra_message_dict["exc_info"] = record.exc_text
+        if record.stack_info:
+            extra_message_dict["stack_info"] = self.formatStack(
+                record.stack_info
+            )
+        extra = record.__dict__.get("extra")
+        if extra:
+            extra_message_dict["extra"] = extra
```

Comment:
One small thing:
I don't think we want the extra object to look like `{"extra": { "some_field": "some_val"}}` - I think we want it to look like `{"some_field": "some_val"}`

which would make this part look like:
```suggestion
        extra = record.__dict__.get("extra")
        if extra and isinstance(extra, dict):
            extra_message_dict.update(extra)
```

Reply:
Makes sense.